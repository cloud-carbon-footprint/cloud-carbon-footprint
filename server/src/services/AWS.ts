import { CloudWatch, CostExplorer } from 'aws-sdk'
import { AWS_REGIONS } from './AWSRegions'

export async function getCostAndUsageResponses(
  params: CostExplorer.GetCostAndUsageRequest,
): Promise<CostExplorer.GetCostAndUsageResponse[]> {
  const costExplorer = new CostExplorer({
    region: AWS_REGIONS.US_EAST_1, //must be us-east-1 to work
  })

  let response: CostExplorer.GetCostAndUsageResponse = {}
  const responses: CostExplorer.GetCostAndUsageResponse[] = []

  do {
    response = await costExplorer.getCostAndUsage({ ...params, NextPageToken: response.NextPageToken }).promise()
    responses.push(response)
  } while (response.NextPageToken)
  return responses
}

export async function getMetricDataResponses(
  params: CloudWatch.GetMetricDataInput,
): Promise<CloudWatch.GetMetricDataOutput[]> {
  const cloudWatch = new CloudWatch()

  let response: CloudWatch.GetMetricDataOutput = {}
  const responses: CloudWatch.GetMetricDataOutput[] = []

  do {
    response = await cloudWatch.getMetricData({ ...params, NextToken: response.NextToken }).promise()
    responses.push(response)
  } while (response.NextToken)

  return responses
}
