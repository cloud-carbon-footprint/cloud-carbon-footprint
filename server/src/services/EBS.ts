import ICloudService from '@domain/ICloudService'
import {
  DiskType,
  getEstimatesFromCostExplorer,
  getUsageFromCostExplorer,
  VolumeUsage,
} from '@services/StorageUsageMapper'
import FootprintEstimate from '@domain/FootprintEstimate'

export default class EBS implements ICloudService {
  serviceName = 'ebs'

  async getEstimates(start: Date, end: Date, region: string): Promise<FootprintEstimate[]> {
    const usage: VolumeUsage[] = await this.getUsage(start, end, region)
    return getEstimatesFromCostExplorer(start, end, region, usage)
  }

  async getUsage(startDate: Date, endDate: Date, region: string): Promise<VolumeUsage[]> {
    const params = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: [
                'EC2: EBS - SSD(gp2)',
                'EC2: EBS - SSD(io1)',
                'EC2: EBS - HDD(sc1)',
                'EC2: EBS - HDD(st1)',
                'EC2: EBS - Magnetic',
              ],
            },
          },
          { Dimensions: { Key: 'REGION', Values: [region] } },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['UsageQuantity'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return await getUsageFromCostExplorer(params, this.getDiskType, region)
  }

  private getDiskType = (awsGroupKey: string) => {
    if (awsGroupKey.endsWith('VolumeUsage.gp2') || awsGroupKey.endsWith('VolumeUsage.piops')) return DiskType.SSD
    if (
      awsGroupKey.endsWith('VolumeUsage.st1') ||
      awsGroupKey.endsWith('VolumeUsage.sc1') ||
      awsGroupKey.endsWith('VolumeUsage')
    )
      return DiskType.HDD
    console.warn('Unexpected Cost explorer Dimension Name: ' + awsGroupKey)
  }
}
