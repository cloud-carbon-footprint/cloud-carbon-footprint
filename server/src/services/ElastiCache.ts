import { config } from 'aws-sdk'
import ComputeUsage from '@domain/ComputeUsage'
import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import { getComputeUsage } from '@services/ComputeUsageMapper'
import { CACHE_NODE_TYPES } from '@services/AWSInstanceTypes'
import { AwsDecorator } from '@services/AwsDecorator'

export default class ElastiCache extends ServiceWithCPUUtilization {
  serviceName = 'elasticache'
  readonly aws: AwsDecorator

  constructor() {
    super()
    this.aws = new AwsDecorator()
  }

  async getUsage(startDate: Date, endDate: Date): Promise<ComputeUsage[]> {
    const cloudWatchParams = {
      StartTime: startDate,
      EndTime: endDate,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/ElastiCache} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const metricDataResponses = await this.aws.getMetricDataResponses(cloudWatchParams)

    const costExplorerParams = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'USAGE_TYPE_GROUP', Values: ['ElastiCache: Running Hours'] } },
          { Dimensions: { Key: 'REGION', Values: [config.region] } },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['UsageQuantity'],
    }
    const costAndUsageResponses = await this.aws.getCostAndUsageResponses(costExplorerParams)

    return getComputeUsage(metricDataResponses, costAndUsageResponses, CACHE_NODE_TYPES)
  }
}
