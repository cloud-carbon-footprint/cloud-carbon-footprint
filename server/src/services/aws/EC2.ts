/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import ComputeUsage, { buildComputeUsages, extractRawComputeUsages, RawComputeUsage } from '@domain/ComputeUsage'
import Cost from '@domain/Cost'
import { GetCostAndUsageRequest } from 'aws-sdk/clients/costexplorer'
import { getCostFromCostExplorer } from '@services/aws/CostMapper'
import { MetricDataResult } from 'aws-sdk/clients/cloudwatch'
import { ServiceWrapper } from '@services/aws/ServiceWrapper'

export default class EC2 extends ServiceWithCPUUtilization {
  serviceName = 'ec2'

  constructor(private serviceWrapper: ServiceWrapper) {
    super()
  }

  async getUsage(start: Date, end: Date): Promise<ComputeUsage[]> {

    const response = await this.serviceWrapper.getQueryByInterval(30, this.runQuery.bind(this), start, end)
    const flattenedResp = response.reduce((acc, data) => [...data, ...acc], [])
    return flattenedResp
  }

  private async runQuery(start: Date, end: Date): Promise<ComputeUsage[]>{
    const params = {
      StartTime: start,
      EndTime: end,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
        {
          Id: 'vCPUs',
          Expression:
            'SEARCH(\'{AWS/Usage,Resource,Type,Service,Class } Resource="vCPU" MetricName="ResourceCount"\', \'Average\', 3600)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const responses = await this.serviceWrapper.getMetricDataResponses(params)

    const metricDataResults: MetricDataResult[] = responses.flatMap((response) => response.MetricDataResults)
    
    const rawComputeUsages: RawComputeUsage[] = metricDataResults.flatMap(extractRawComputeUsages)
    return buildComputeUsages(rawComputeUsages, 'AWS')
  }

  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {

    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'REGION',
              Values: [region],
            },
          },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['EC2: Running Hours'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['AmortizedCost'],
    }
    return getCostFromCostExplorer(params, this.serviceWrapper)
  }
}
