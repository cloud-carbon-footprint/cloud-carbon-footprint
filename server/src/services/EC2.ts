import AWS from 'aws-sdk'
import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import ComputeUsage from '@domain/ComputeUsage'

export default class EC2 extends ServiceWithCPUUtilization {
  serviceName = 'ec2'
  readonly cloudWatch: AWS.CloudWatch

  constructor() {
    super()
    this.cloudWatch = new AWS.CloudWatch()
  }

  async getUsage(startDate: Date, endDate: Date): Promise<ComputeUsage[]> {
    const params = {
      StartTime: startDate,
      EndTime: endDate,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
        {
          Id: 'vCPUs',
          Expression:
            'SEARCH(\'{AWS/Usage,Resource,Type,Service,Class } Resource="vCPU" MetricName="ResourceCount"\', \'Average\', 3600)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const response = await this.cloudWatch.getMetricData(params).promise()

    interface RawComputeUsage {
      cpuUtilization?: number[]
      vCPUCount?: number
      timestamp?: Date
      cpuUtilizationAvg?: number
    }
    const result: { [key: string]: RawComputeUsage } = {}

    // Aggregate CPU Utilization
    const metricDataResults = response.MetricDataResults
    const cpuUtilizationData = metricDataResults.filter((a) => a.Id === 'cpuUtilization')
    cpuUtilizationData.forEach((instanceCPUUtilization) => {
      instanceCPUUtilization.Timestamps.forEach((timestamp, i) => {
        const key = new Date(timestamp).toISOString()
        if (!result[key]) result[key] = { cpuUtilization: [instanceCPUUtilization.Values[i]], timestamp: timestamp }
        result[key].cpuUtilization.push()
      })
    })

    // Add vCPU to result object
    const vCPUData = metricDataResults.filter((a) => a.Id === 'vCPUs')
    vCPUData[0]?.Timestamps.forEach((timestamp, i) => {
      const key = new Date(timestamp).toISOString()
      if (!result[key]) result[key] = { cpuUtilization: [], timestamp: timestamp }
      result[key].vCPUCount = vCPUData[0].Values[i]
    })

    // Apply estimation formula
    Object.values(result).forEach((a) => {
      a.cpuUtilizationAvg =
        a.cpuUtilization.reduce((acc, a) => {
          return acc + a
        }, 0) / a.cpuUtilization.length

      if (isNaN(a.cpuUtilizationAvg)) {
        a.cpuUtilizationAvg = 0.0
      }
    })

    // Aggregate by date
    const estimationsByDay: { [timestamp: string]: ComputeUsage } = {}
    Object.values(result).forEach((a) => {
      const timestamp = new Date(a.timestamp)
      const date = timestamp.toISOString().substr(0, 10)

      if (!estimationsByDay[date]) {
        estimationsByDay[date] = {
          timestamp: new Date(date),
          cpuUtilizationAverage: a.cpuUtilizationAvg,
          numberOfvCpus: a.vCPUCount,
        }
      } else {
        estimationsByDay[date].cpuUtilizationAverage += a.cpuUtilizationAvg
        estimationsByDay[date].numberOfvCpus += a.vCPUCount
      }
    })

    return Object.values(estimationsByDay)
  }
}
