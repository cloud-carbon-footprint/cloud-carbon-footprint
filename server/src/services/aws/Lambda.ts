import ICloudService from '@domain/ICloudService'
import FootprintEstimate from '@domain/FootprintEstimate'
import { estimateCo2, CLOUD_CONSTANTS } from '@domain/FootprintEstimationConstants'
import { getCostFromCostExplorer } from '@services/aws/CostMapper'
import Cost from '@domain/Cost'
import { isEmpty } from 'ramda'
import { GetCostAndUsageRequest } from 'aws-sdk/clients/costexplorer'
import { CloudWatchLogs } from 'aws-sdk'
import { ServiceWrapper } from '@services/aws/ServiceWrapper'

export default class Lambda implements ICloudService {
  serviceName = 'lambda'

  constructor(
    private TIMEOUT = 60000,
    private POLL_INTERVAL = 1000,
    private readonly cloudWatchLogs: CloudWatchLogs,
    private readonly serviceWrapper: ServiceWrapper,
  ) {}

  async getEstimates(start: Date, end: Date, region: string): Promise<FootprintEstimate[]> {
    const groupNames = await this.getLambdaLogGroupNames()

    if (isEmpty(groupNames)) {
      return []
    }

    // given start and end date
    // map over each month between start and end date
    // execute runQuery o

    const queryId = await this.runQuery(start, end, groupNames)
    const usage = await this.getResults(queryId)

    return usage.results.map((resultByDate) => {
      const timestampField = resultByDate[0]
      const wattsField = resultByDate[1]
      const timestamp = new Date(timestampField.value.substr(0, 10))
      const wattHours = Number.parseFloat(wattsField.value)
      const co2e = estimateCo2(wattHours, region)
      return {
        timestamp,
        wattHours,
        co2e,
      }
    })
  }

  private async getLambdaLogGroupNames(): Promise<string[]> {
    const params = {
      logGroupNamePrefix: '/aws/lambda',
    }

    const logGroupData = await this.cloudWatchLogs.describeLogGroups(params).promise()
    return logGroupData.logGroups.map(({ logGroupName }) => logGroupName)
  }

  private async runQuery(start: Date, end: Date, groupNames: string[]): Promise<string> {
    const query = `
            filter @type = "REPORT"
            | fields datefloor(@timestamp, 1d) as Date, @duration/1000 as DurationInS, @memorySize/1000000 as MemorySetInMB, ${CLOUD_CONSTANTS.AWS.MAX_WATTS} * DurationInS/3600 * MemorySetInMB/1792 as wattsPerFunction
            | stats sum(wattsPerFunction) as Watts by Date 
            | sort Date asc`

    const params = {
      startTime: start.getTime(),
      endTime: end.getTime(),
      queryString: query,
      logGroupNames: groupNames,
    }
    const queryData = await this.cloudWatchLogs.startQuery(params).promise()
    return queryData.queryId
  }

  private async getResults(queryId: string) {
    const params = {
      queryId: queryId,
    }
    let cwResultsData
    const startTime = Date.now()

    while (true) {
      cwResultsData = await this.cloudWatchLogs.getQueryResults(params).promise()
      if (cwResultsData.status !== 'Running' && cwResultsData.status !== 'Scheduled') break
      if (Date.now() - startTime > this.TIMEOUT) {
        throw new Error(`CloudWatchLog request failed, status: ${cwResultsData.status}`)
      }
      await wait(this.POLL_INTERVAL)
    }
    return cwResultsData
  }

  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'REGION',
              Values: [region],
            },
          },
          {
            Dimensions: {
              Key: 'SERVICE',
              Values: ['AWS Lambda'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['AmortizedCost'],
    }
    return getCostFromCostExplorer(params, this.serviceWrapper)
  }
}

async function wait(ms: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}
