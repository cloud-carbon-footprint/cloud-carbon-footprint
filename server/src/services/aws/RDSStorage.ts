import { GetCostAndUsageRequest } from 'aws-sdk/clients/costexplorer'
import ICloudService from '@domain/ICloudService'
import {
  DiskType,
  getEstimatesFromCostExplorer,
  getUsageFromCostExplorer,
  VolumeUsage,
} from '@services/aws/StorageUsageMapper'
import FootprintEstimate from '@domain/FootprintEstimate'
import Cost from '@domain/Cost'
import { getCostFromCostExplorer } from '@services/aws/CostMapper'
import { ServiceWrapper } from '@services/aws/ServiceWrapper'

export default class RDSStorage implements ICloudService {
  serviceName = 'rds-storage'

  constructor(private readonly serviceWrapper: ServiceWrapper) {}

  async getEstimates(start: Date, end: Date, region: string): Promise<FootprintEstimate[]> {
    const usage: VolumeUsage[] = await this.getUsage(start, end, region)
    return getEstimatesFromCostExplorer(start, end, region, usage)
  }

  async getUsage(startDate: Date, endDate: Date, region: string): Promise<VolumeUsage[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'REGION', Values: [region] } },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['RDS: Storage'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['UsageQuantity'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return await getUsageFromCostExplorer(params, this.getDiskType, this.serviceWrapper)
  }

  private getDiskType = (awsGroupKey: string) => {
    if (awsGroupKey.endsWith('GP2-Storage') || awsGroupKey.endsWith('PIOPS-Storage')) return DiskType.SSD
    if (awsGroupKey.endsWith('StorageUsage')) return DiskType.HDD
    console.warn('Unexpected Cost explorer Dimension Name: ' + awsGroupKey)
  }

  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'REGION', Values: [region] } },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['RDS: Storage'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['AmortizedCost'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return getCostFromCostExplorer(params, this.serviceWrapper)
  }
}
