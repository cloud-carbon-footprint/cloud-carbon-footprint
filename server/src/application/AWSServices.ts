import CloudService from '@domain/CloudService'
import EBS from '@services/EBS'
import S3 from '@services/S3'
import EC2 from '@services/EC2'
import ElastiCache from '@services/ElastiCache'
import RDSComputeService from '@services/RDSCompute'

export default function AWSServices(): CloudService[] {
  return [new EBS(), new S3(), new EC2(), new ElastiCache(), new RDSComputeService()]
}
