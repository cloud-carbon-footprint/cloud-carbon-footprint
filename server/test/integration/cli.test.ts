import cli from '@view/cli'
import AWSServices from '@application/AWSServices'
import RDS from '@services/RDS'
import RDSStorage from '@services/RDSStorage'
import RDSComputeService from '@services/RDSCompute'
import EBS from '@services/EBS'
import S3 from '@services/S3'
import EC2 from '@services/EC2'
import ElastiCache from '@services/ElastiCache'
import AWSMock from 'aws-sdk-mock'
import AWS from 'aws-sdk'
import { mocked } from 'ts-jest/utils'
import Lambda from '@services/Lambda'
import {
  mockAwsCloudWatchGetMetricData,
  mockAwsCloudWatchGetQueryResultsForLambda,
  mockAwsCostExplorerGetCostAndUsage,
} from '../fixtures/awsMockFunctions'

jest.mock('@application/AWSServices')

beforeAll(() => {
  AWSMock.setSDKInstance(AWS)
})

afterEach(() => {
  AWSMock.restore()
})

describe('cli', () => {
  const start = '2020-07-01'
  const end = '2020-07-07'
  const servicesRegistered = mocked(AWSServices, true)
  const rawRequest = ['executable', 'file', '--startDate', start, '--endDate', end, '--region', 'us-east-1']

  describe('ebs, s3, ec3, elasticache, rds', () => {
    servicesRegistered.mockReturnValue([
      new EBS(),
      new S3(),
      new EC2(),
      new ElastiCache(),
      new RDS(new RDSComputeService(), new RDSStorage()),
    ])

    test('ebs, s3, ec2, elasticache, rds, grouped by day and service', async () => {
      mockAwsCloudWatchGetMetricData()
      mockAwsCostExplorerGetCostAndUsage()

      const result = await cli(rawRequest)

      expect(result).toMatchSnapshot()
    })

    test('ebs, s3, ec2, elasticache, rds grouped by service', async () => {
      mockAwsCloudWatchGetMetricData()
      mockAwsCostExplorerGetCostAndUsage()

      const result = await cli([...rawRequest, '--groupBy', 'service'])

      expect(result).toMatchSnapshot()
    })

    test('ebs, s3, ec2, elasticache, rds, grouped by day', async () => {
      mockAwsCloudWatchGetMetricData()
      mockAwsCostExplorerGetCostAndUsage()

      const result = await cli([...rawRequest, '--groupBy', 'day'])

      expect(result).toMatchSnapshot()
    })
  })

  describe('lambda', () => {
    it('lambda cli test', async () => {
      servicesRegistered.mockReturnValue([new Lambda()])
      mockAwsCloudWatchGetQueryResultsForLambda()

      const result = await cli([...rawRequest, '--groupBy', 'dayAndService'])

      expect(result).toMatchSnapshot()
    })
  })
})
