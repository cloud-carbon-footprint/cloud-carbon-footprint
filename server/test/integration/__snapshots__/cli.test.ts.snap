// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds grouped by service 1`] = `
"| Service      | Watt Hours        | Kg CO2e Emissions      | Cost              
| EBS          | 3.21              | 0.001083               | $4.62             
| S3           | 0.05              | 0.000017               | $3.00             
| EC2          | 18.81             | 0.006337               | $0.00             
| ElastiCache  | 2.34              | 0.000787               | $3.00             
| RDS          | 4.98              | 0.001679               | $9.23             
| Lambda       | 0.00              | 0.000000               | $0.00             
| Total        | 29.39             | 0.009903               | $19.85            "
`;

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds, grouped by day 1`] = `
"| Date         | Watt Hours        | Kg CO2e Emissions      
| 2020-07-01   | 26.18             | 0.008820               
| 2020-07-02   | 3.21              | 0.001083               
| Total        | 29.39             | 0.009903               "
`;

exports[`cli ebs, s3, ec3, elasticache, rds ebs, s3, ec2, elasticache, rds, grouped by day and service 1`] = `
"| Date (UTC)   | EBS Watt Hours    | EBS Kg CO2e Emissions  | S3 Watt Hours     | S3 Kg CO2e Emissions   | EC2 Watt Hours    | EC2 Kg CO2e Emissions  | ElastiCache Watt Hours| ElastiCache Kg CO2e Emissions| RDS Watt Hours    | RDS Kg CO2e Emissions  | Lambda Watt Hours | Lambda Kg CO2e Emissions| SUM Watt Hours    | SUM Kg CO2e Emissions  
| 2020-07-01   | 1.07              | 0.000361               | 0.05              | 0.000017               | 18.81             | 0.006337               | 2.34              | 0.000787               | 3.91              | 0.001318               | 0.00              | 0.000000               | 26.18             | 0.008820               
| 2020-07-02   | 2.14              | 0.000722               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 1.07              | 0.000361               | 0.00              | 0.000000               | 3.21              | 0.001083               
| Total        | 3.21              | 0.001083               | 0.05              | 0.000017               | 18.81             | 0.006337               | 2.34              | 0.000787               | 4.98              | 0.001679               | 0.00              | 0.000000               | 29.39             | 0.009903               "
`;

exports[`cli lambda lambda cost 1`] = `
"| Service      | Watt Hours        | Kg CO2e Emissions      | Cost              
| EBS          | 0.00              | 0.000000               | $0.00             
| S3           | 0.00              | 0.000000               | $0.00             
| EC2          | 0.00              | 0.000000               | $0.00             
| ElastiCache  | 0.00              | 0.000000               | $0.00             
| RDS          | 0.00              | 0.000000               | $0.00             
| Lambda       | 0.43              | 0.000145               | $120.00           
| Total        | 0.43              | 0.000145               | $120.00           "
`;

exports[`cli lambda lambda estimates 1`] = `
"| Date (UTC)   | EBS Watt Hours    | EBS Kg CO2e Emissions  | S3 Watt Hours     | S3 Kg CO2e Emissions   | EC2 Watt Hours    | EC2 Kg CO2e Emissions  | ElastiCache Watt Hours| ElastiCache Kg CO2e Emissions| RDS Watt Hours    | RDS Kg CO2e Emissions  | Lambda Watt Hours | Lambda Kg CO2e Emissions| SUM Watt Hours    | SUM Kg CO2e Emissions  
| 2020-07-01   | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.20              | 0.000067               | 0.20              | 0.000067               
| 2020-07-02   | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.23              | 0.000077               | 0.23              | 0.000077               
| Total        | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.00              | 0.000000               | 0.43              | 0.000145               | 0.43              | 0.000145               "
`;
